pkg/gitops/audit_handler.go

```diff
@@ -14,6 +14,82 @@
 
 package gitops
 
+import (
+	"fmt"
+	"bytes"
+    "errors"
+	"encoding/json"
+	"k8s.io/klog/v2"
+	auditv1 "k8s.io/apiserver/pkg/apis/audit/v1"
+)
+
 func (cr *CustomRepo) HandleEventList(jsonstring []byte) error {
+	cr.Mutex.Lock()
+	defer cr.Mutex.Unlock()
+	eventList := auditv1.EventList{}
+	jsonstring = bytes.TrimPrefix(jsonstring, []byte("\xef\xbb\xbf"))
+	err := json.Unmarshal(jsonstring, &eventList)
+	if err != nil {
+		klog.ErrorS(err, "unable to unmarshal json into event list struct")
+		return err
+	}
+	for _, event := range eventList.Items {
+		if event.Stage != "ResponseComplete" ||
+			event.ResponseStatus.Status == "Failure" ||
+			event.User.Username == cr.ServiceAccount {
+			klog.V(2).InfoS("Audit event skipped (audit Stage != ResponseComplete, audit ResponseStatus != Success, or audit produced by rollback)")
+			continue
+		}
+		if cr.RollbackMode {
+			klog.V(2).ErrorS(err, "could not handle event with rollback in progress")
+			return fmt.Errorf("rollback in progress")
+		}
+        if err = cr.HandleEvent(event); err != nil {
+			klog.V(2).ErrorS(err, "could not handle event")
+            return err
+        }
+	}
 	return nil
 }
+
+func (cr *CustomRepo) HandleEvent(event auditv1.Event) (error) {
+    user := event.User.Username
+    email := event.User.Username + "+" + event.User.UID + "@audit.antrea.io"
+	message := resourceMap[event.ObjectRef.Resource+event.ObjectRef.APIGroup] + event.ObjectRef.Namespace + "/" + event.ObjectRef.Name
+	switch verb := event.Verb; verb {
+	case "create":
+		if err := cr.modifyFile(event); err != nil {
+			klog.V(2).ErrorS(err, "could not create new resource")
+			return err
+		}
+		if err := cr.AddAndCommit(user, email, "Created "+message); err != nil {
+			klog.V(2).ErrorS(err, "could not add/commit add operation")
+			return err
```

Comment:
same general remark as for @stanleywbwong's earlier PRs. If you are going to return an error and log it further up the call stack, you don't need to log it here as well. It can be very verbose and you also end up with multiple log entries in the log file for the same error. Typically you want code that looks like this:

```go
func A() {
    if err := B(); err != nil {
        klog.ErrorS(err, "failed to do that")
    }
}

func B() error {
    if err := C(); err != nil {
        return fmt.Errorf("more context and error wrapping: %w", err)
    }
}

func C() error {
    return fmt.Errorf("operation Foo failed")
}
```

If it's an error, don't use `V(2)` (I'm actually surprised there is more than 1 verbosity for errors)

Reply:
in that case, would the only klog.ErrorS() statements be found in the main.go files where the top-level calls happen?

Reply:
Either that or the top level gitops functions. Which would better for our case?

Reply:
I'll leave them in the main.go files and return errors from everywhere else for now.

Reply:
yes typically that means propagating an error to a "top-level" function. But, sometimes you need to log an error earlier without returning the error. For example, if I have some retry mechanism:

```go
func A() {
    if err := B(); err != nil {
        klog.ErrorS(err, "failed to do that")
    }
}

func B() error {
    if ... {
        return fmt.Errorf("more context and error wrapping: %w", err)
    }
    // this task (C) will be attempted asynchronously until it succeeds
    go func() {
        const retryInterval = 1*time.Second
        for {
            if err := C(); err != nil {
                klog.ErrorS(err, "could not do that", "retryInterval", retryInterval)
            } else {
                return // success
            }
        }
    }
    return nil
}

func C() error {
    return fmt.Errorf("operation Foo failed")
}
```

a "top-level" function can also be many different things (could be main, could be an HTTP handler, etc.)